src/init.c

```diff
@@ -345,8 +345,15 @@ int uv_dup(uv_os_fd_t fd, uv_os_fd_t* dupfd) {
 }
 #else
 int uv_dup(uv_os_fd_t fd, uv_os_fd_t* dupfd) {
+// F_DUPFD_CLOEXEC only available since Linux 2.6.24
+#ifdef F_DUPFD_CLOEXEC
     if ((*dupfd = fcntl(fd, F_DUPFD_CLOEXEC, 3)) == -1)
         return -errno;
+#else
+    if ((*dupfd = fcntl(fd, F_DUPFD, 3)) == -1)
```

Comment:
This should be unconditionally compiled and be used if the `F_DUPFD_CLOEXEC` one fails. In another word, this is only handling the compile time kernel header version, not the runtime kernel version. The generic binary, for example, are likely going to be used on a different kernel than the one it is compiled on.

Reply:
I thought it might be undefined behavior to use `F_DUPFD_CLOEXEC` on older kernels?  Is that okay?

Reply:
Whether that's undefined or not, nothing in this PR can stop that. The change I request only make it better if it wasn't undefined.

And in general, I believe the kernel API are written so that passing a newer flag to an older kernel should be fine. You'll just get a `EINVAL` if you do that. The kernel does all the checks to make sure the flag you pass in is what it understand before it proceed. It should even be fine to define that value manually when compiling with an older header. It just might be arch dependent (it's not syscall number so hopefully not but one need to check) and generally doesn't worth the effort unless the buildbot has an old kernel. See also `fcntl(3p)`

>        EINVAL The cmd argument is invalid, or the cmd argument is F_DUPFD or F_DUPFD_CLOEXEC  and
>              arg  is  negative  or  greater  than or equal to {OPEN_MAX}, or the cmd argument is
>              F_GETLK, F_SETLK, or F_SETLKW and the data pointed to  by  arg  is  not  valid,  or
>              fildes refers to a file that does not support locking.

FWIW, I think UB is mostly a C thing. I don't think the kernel is in that business probably mostly for security and backward compatibility reasons....

Reply:
Yeah, there's a more complete version of this as https://github.com/libuv/libuv/pull/1994